src/hotspot/share/gc/z/zObjArrayAllocator.cpp

```diff
@@ -50,9 +50,18 @@ oop ZObjArrayAllocator::initialize(HeapWord* mem) const {
   // time and time-to-safepoint
   const size_t segment_max = ZUtils::bytes_to_words(64 * K);
   const BasicType element_type = ArrayKlass::cast(_klass)->element_type();
-  const size_t header = arrayOopDesc::header_size(element_type);
-  const size_t payload_size = _word_size - header;
+  int base_offset = arrayOopDesc::base_offset_in_bytes(element_type);
+
+  // Clear leading 32 bits, if necessary.
+  if (!is_aligned(base_offset, HeapWordSize)) {
+    assert(is_aligned(base_offset, BytesPerInt), "array base must be 32 bit aligned");
+    *reinterpret_cast<jint*>(reinterpret_cast<char*>(mem) + base_offset) = 0;
+    base_offset += BytesPerInt;
+  }
+  assert(is_aligned(base_offset, HeapWordSize), "remaining array base must be 64 bit aligned");
 
+  const size_t header = heap_word_size(base_offset);
```

Comment:
`base_offset` has already been aligned by the code above, so we don't have to call heap_word_size here and can simply divide with HeapWordSize.

Reply:
Regarding MemAllocator::obj_memory_range(): why do we even need to exclude array header from the bad-heap-word-check, when not zeroing? It seems reasonable to also check the header word(s) in any case. I'd just remove ObjArrayAllocator::obj_memory_range() and always use the full object range, as implemented by MemAllocator::obj_memory_range().